Application profiling

marcins opened the pull request:
<!---
Thanks for filing a pull request 😄 ! Before you submit, please read the following:

Search open/closed issues before submitting since someone might have pushed the same thing before!
-->

# ↪️ Pull Request

<!---
Provide a general summary of the pull request here
Please look for any issues that this PR resolves and tag them in the PR.
-->

This is an implementation of the Tracing RFC - more detail on the background and motivations is in #8588. The main differences from the RFC is that the feature is called "application profiling" (and the existing CPU profiling is referred to as "Sampling Profiling").

Instrumentation has been added for:
* high-level phases - building, bundling, packaging
* transform (per-plugin)
* Babel transform (per-babel plugin visitor)
* bundling (bundle / optimize)
* packaging (package / optimize)
* other plugin types (resolving, naming, etc..)

## 🚨 Test instructions

An app profile can be generated by starting Parcel with `--profile-application`.

## ✔️ PR Todo

- [x] Inject the profiler for plugins
- [x] Create a low-overhead instrumentation method
- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

See also documentation PR: https://github.com/parcel-bundler/website/pull/1075

marcins committed:
Move profiler to new @parcel/profiler pkg

marcins committed:
Refactor Profiler -> SamplingProfiler

marcins committed:
Initial implementation of ApplicationProfiler

marcins committed:
Plumb in applicationProfiler to child/worker/reporter system

marcins committed:
Add initial application profile reporter

marcins committed:
Implement CLI integration for application profiler

marcins committed:
Implement profiling for packaging

marcins committed:
Implement profiling for bundling

marcins committed:
Implement enabling app profiling in workers

marcins committed:
Implement profiling for Babel transforms

marcins committed:
Fix race condition with closing trace writeStream

marcins committed:
Implement profiling for top level phases

marcins committed:
Implement profiling for resolution

marcins committed:
Implement profiling for optimization

marcins committed:
Implement profiling for bundle naming

mischnic commented on packages/transformers/babel/src/babel7.js:
```diff
@@ -8,6 +8,8 @@ import {relativeUrl} from '@parcel/utils';
 import {remapAstLocations} from './remapAstLocations';
 
 import packageJson from '../package.json';
+import {applicationProfiler} from '@parcel/profiler';
```
I think it be better to have this be provided by core (so "dependency injection" similar to the `logger` right now).

This would break if Yarn doesn't hoist the package and then the calls `applicationProfiler.onTrace()` and `applicationProfiler.createMeasurement` would actually go to different instances.

We had this problem in the past: https://github.com/parcel-bundler/parcel/pull/6566

marcins replied:
Yep, that's fair enough - you mean just where it's used in plugins as opposed to core? 

FWIW the initial implementation that Brian did internally passed around a "tracer" through core as well - this worked okay for the top level metrics, but once you wanted to instrument anything deeper it was a real pain to thread it through all the various layers.

mischnic replied:
It's only a problem if some package X has state (such as that profiler) and X is declared as a dependency in multiple packages. 
So just importing it inside core would be fine (as there's only one dependency and no way Yarn creates two instances), though if it's only used inside of core anyway then it could maybe also just be separate file in core (as opposed to a separate package).

mischnic replied:
> you mean [using dependency injection] just where it's used in plugins as opposed to core?

Yes, in the public API

lettertwo commented on packages/core/core/src/PackagerRunner.js:
```diff
@@ -394,7 +395,15 @@ export default class PackagerRunner {
     let {name, resolveFrom, plugin} = packager;
+    let measurement;
     try {
+      measurement = applicationProfiler.createMeasurement(name, {
```
Should this be gated by a check of `applicationProfiler.enabled`? Alternatively, could `applicationProfiler` expose category-specific methods that take positional args (instead of requiring object allocations) that are no-op by default?

marcins replied:
Yeah, good idea - given most of the calls follow the same pattern with just name, category and the filename changing can create a function for that use case with 3 string args that exits early if disabled (or as you say, change all of these to `applicationProfiler.enabled && applicationProfiler.createMeasurement(..)` - that would make `measurement` falsy which means later the `measurement && measurement.end()` would work.

lettertwo commented on packages/core/core/src/PackagerRunner.js:
```diff
@@ -495,7 +505,14 @@ export default class PackagerRunner {
     for (let optimizer of optimizers) {
+      let measurement;
       try {
+        measurement = applicationProfiler.createMeasurement(optimizer.name, {
```
Similarly, should this be gated or otherwise made to avoid the object allocations?

lettertwo commented on packages/core/core/src/Transformation.js:
```diff
@@ -435,6 +436,16 @@ export default class Transformation {
     let resultingAssets = [];
     let finalAssets = [];
     for (let transformer of pipeline.transformers) {
+      let measurement = applicationProfiler.createMeasurement(
```
Here also? I am thinking that in the default case (profiling is not enabled), calls to `createMeasurement` should be a no-op, including skipping the allocation of new argument bags. I'm aware that this feels a bit like premature optimization; reasonable folks can disagree.

marcins replied:
I haven't done any significant perf measurement, but given most "prod" builds would be running with the profiler disabled avoiding unnecessary allocations / GC overhead seems like a reasonable idea.

lettertwo commented on packages/core/core/src/requests/PathRequest.js:
```diff
@@ -194,7 +195,14 @@ export class ResolverRunner {
     for (let resolver of resolvers) {
+      let measurement;
       try {
+        measurement = applicationProfiler.createMeasurement(resolver.name, {
```
Flagging this one as part of the same series of comments. (not sure if i'm gonna catch em all :pokeball:)

marcins replied:
I'll just search the code for all the `createMeasurement()` calls and address this.

lettertwo commented on packages/core/profiler/src/ApplicationProfiler.js:
```diff
@@ -0,0 +1,82 @@
+    name: string,
+    data?: ApplicationProfilerMeasurementData = {categories: ['Core']},
+  ): ApplicationProfilerMeasurement {
+    const start = performance.now();
```
Should this method just no-op if `this.enabled === false`? Alternatively, call sites could be gated by that same check, which isn't a bad idea given that it could avoid allocating the data objects if profiling is disabled.

Another idea (mentioned in my first comment) could be to add category-specific methods that take positional arguments, which could allow call sites to avoid creating the complex `data` object. Those methods could just no-op by default, or build the args and call `createMeasurement` otherwise.

mattcompiles commented on packages/core/core/src/requests/ParcelBuildRequest.js:
```diff
@@ -89,12 +94,14 @@ async function run({input, api, options}: RunInput) {
     ),
   });
 
+  let packaingMeasurement = applicationProfiler.createMeasurement('packaging');
```
```suggestion
  let packagingMeasurement = applicationProfiler.createMeasurement('packaging');
```

mattcompiles commented on packages/core/core/src/requests/ParcelBuildRequest.js:
```diff
@@ -89,12 +94,14 @@ async function run({input, api, options}: RunInput) {
   });
 
   let bundleInfo = await api.runRequest(writeBundlesRequest);
+  packaingMeasurement.end();
```
```suggestion
  packagingMeasurement.end();
```

marcins committed:
Bump profiler to 2.8.3 to match rest

marcins committed:
Refactor to use PluginApplicationProfiler

marcins committed:
Refactor createMeasurement to avoid redundant obj allocations

marcins committed:
Document PluginApplicationProfiler API

marcins committed:
Implement injection of PluginApplicationProfiler

marcins committed:
Implement reporting for reporters..

marcins committed:
Implement application profiling for applyRuntime

marcins committed:
Add application profiling to compressor

marcins committed:
Fix an errant relative path

marcins commented:
I've finally gotten to addressing the PR feedback from late last year. The main changes have been:
* rebasing on latest v2 branch
* “injecting” a `PluginApplicationProfiler` similar to `PluginLogger` for plugins to ensure the same instance is used
instrumenting the remaining plugin types in core

What’s still missing:
* adding some test coverage to the new functionality

mischnic commented on packages/core/types/index.js:
```diff
@@ -1931,3 +1976,31 @@ export interface IDisposable {
+   * of arguments to `createMeasurement` - for example if you need to determine the entry of a bundle to pass it
+   * in as the <code>argumentName</code>, you would only do this if the application profiler is enabled.
+   */
+  get enabled(): boolean;
```
Nit: for all the other public types, we've been using a readonly property (so `+enabled: boolean;`). For example `bundle.needsStableName` is in reality a getter in `core/src/public/Bundle.js`

mischnic commented on packages/core/types/index.js:
```diff
@@ -1931,3 +1976,31 @@ export interface IDisposable {
+   *
+   * <code>otherArgs</code> can be used for specifying any other key/value pairs that should be written to the profile.
+   *
+   * For example: <code>profiler.createMeasurement('compiled', 'find_css', fromProjectPathRelative(asset.filePath), { meta: 'data' })</code>
```
`fromProjectPathRelative` is an internal API. So either `asset.filePath` or `path.relative(options.projectRoot, asset.filePath)`

mischnic commented on packages/reporters/application-profiler/src/ApplicationProfilerReporter.js:
```diff
@@ -0,0 +1,85 @@
+import {Reporter} from '@parcel/plugin';
+import {Tracer} from 'chrome-trace-event';
+
+// We need to maintain some state here to ensure we write to the same output, there should only be one
```
Sidenote: This is exactly why `options.instanceId` was added. But in reality, this probably never happens anyway

mischnic commented on packages/transformers/babel/src/babel7.js:
```diff
@@ -66,6 +79,27 @@ export default async function babel7(
+    ) => {
+      return function () {
+        let measurement;
+        if (applicationProfiler.enabled) {
```
Can the profiler be disabled mid-build or is the inner check unnecessary?

marcins replied:
Ah yep, I can remove this inner check.

mischnic commented on packages/core/parcel/src/cli.js:
```diff
@@ -464,6 +465,13 @@ async function normalizeOptions(
+  if (command.profileApplication) {
+    additionalReporters.unshift({
+      packageName: '@parcel/reporter-application-profiler',
+      resolveFrom: __filename,
```
This means that `@parcel/reporter-application-profiler` needs to be a dependency of the `parcel` CLI.

mischnic commented on packages/core/core/src/requests/PathRequest.js:
```diff
@@ -375,17 +387,7 @@ export class ResolverRunner {
 
         break;
       } finally {
-        // Add dev dependency for the resolver. This must be done AFTER running it due to
-        // the potential for lazy require() that aren't executed until the request runs.
-        let devDepRequest = await createDevDependency(
-          {
-            specifier: resolver.name,
-            resolveFrom: resolver.resolveFrom,
-          },
-          this.previousDevDeps,
-          this.options,
-        );
-        this.runDevDepRequest(devDepRequest);
```
Looks like an error during rebasing

mischnic commented on packages/core/parcel/package.json:
```diff
@@ -14,7 +14,7 @@
     "url": "https://github.com/parcel-bundler/parcel.git"
   },
-  "bin": "src/bin.js",
+  "bin": "lib/bin.js",
```
```suggestion
  "bin": "src/bin.js",
```

marcins replied:
Oops, yes, good catch - I accidentally committed this!

marcins commented:
One thing to add - I did consider at some point whether instead of adding `PluginApplicationProfiler` everywhere I should just add the `createMeasurement` method to the existing `PublicLogger` as a profiling measurement can be argued to be a kind of log. In the end I talked myself out of it to avoid complexity with interdependencies between the logger & profiler, but happy to hear any other feels on this.

marcins committed:
Remove redundant profiler enabled check

marcins committed:
Fix bad rebase

marcins committed:
Fix bad commit to core package.json

marcins committed:
Add app profiler reporter as dependency of CLI

marcins committed:
Fix types / documentation for PluginAppProfiler

mischnic commented on packages/transformers/babel/src/babel7.js:
```diff
@@ -66,6 +79,25 @@ export default async function babel7(
+      fn: Function,
+    ) => {
+      return function () {
+        let measurement;
+        measurement = applicationProfiler.createMeasurement(
```
```suggestion
        let measurement = applicationProfiler.createMeasurement(
```

marcins committed:
Add unit tests for application profiler

marcins committed:
Remove a redundant allocation

marcins committed:
Move transform measurement to better location

marcins committed:
Minor improvements to some measurements

marcins committed:
Prevent double-end of measurement

marcins committed:
Improve profile reporter

marcins committed:
Add basic integration test for app profiling

lettertwo reviewed (approved):
😍

marcins committed:
Merge remote-tracking branch 'upstream/v2' into mszczepanski/application-profiling

marcins committed:
Merge remote-tracking branch 'upstream/v2' into mszczepanski/application-profiling

marcins committed:
Fix linting issues

lettertwo committed:
Merge branch 'v2' into mszczepanski/application-profiling

devongovett commented:
This will be a really useful feature! My only question is around naming. I'm not sure what "application" refers to here. It's not really profiling the user's application so much as it is profiling Parcel. It seems more like a "trace" to me. Maybe we could have `--profile` and `--trace` as flags, and `trace` or `tracer` as the object passed to plugins? What do you think?

lettertwo committed:
Merge branch 'v2' into mszczepanski/application-profiling

lettertwo committed:
Rename ApplicationProfiler -> Tracer

Also rename:
- @parcel/reporter-application-profiler -> @parcel/reporter-tracer
- PluginApplicationProfiler -> PluginTracer
- cli option `--profile-application` -> `--trace`
- PluginOptions.applicationProfiler -> PluginOptions.tracer

lettertwo committed:
Merge branch 'v2' into mszczepanski/application-profiling

lettertwo commented:
Renamed ApplicationProfiler -> Tracer

Also renamed:
- @parcel/reporter-application-profiler -> @parcel/reporter-tracer
- PluginApplicationProfiler -> PluginTracer
- cli option `--profile-application` -> `--trace`
- PluginOptions.applicationProfiler -> PluginOptions.tracer

lettertwo commented:
@devongovett I think the rename is good, but with one small point of confusion: The `SamplingProfiler` has an accompanying `Trace` that is now very similar in name to `Tracer` (formerly called `ApplicationProfiler`.  Should we rename `Trace` to something that ties it to `SamplingProfiler`?

devongovett commented:
That's not exposed externally to parcel though right? If not then doesn't seem like too big a deal.

devongovett commented on packages/core/workers/index.d.ts:
```diff
@@ -11,6 +11,7 @@ export type FarmOptions = {
   workerPath?: FilePath,
   backend: BackendType,
   shouldPatchConsole?: boolean,
+  shouldProfileApplication?: boolean,
```
```suggestion
  shouldTrace?: boolean,
```

devongovett commented on packages/transformers/babel/src/babel7.js:
```diff
@@ -74,6 +76,28 @@ export default async function babel7(
   };
 
+  if (tracer.enabled) {
+    config.wrapPluginVisitorMethod = (
```
Does this wrap visiting each individual AST node, or just once for each plugin? How expensive is that?

marcins replied:
> Does this wrap visiting each individual AST node, or just once for each plugin? How expensive is that?

I just ran two full production Jira builds back-to-back, one with tracing & one without (both with `--no-cache`). The timings were 18m 29s without tracing, and 19m 1s with tracing (on an M1 Mac). So minimal overhead for tracing in general. 

The trace file produced is approximately 1.4GB - can't load it directly in the Perfetto UI (the WASM version), but works fine via the `trace_processor` CLI tool.

There are approximately 2.9m Babel transform entries - so yes, my understanding is that it wraps every visitor call for every AST type, but as you can see the overhead of this is relatively minimal.

marcins committed:
Merge remote-tracking branch 'upstream/v2' into mszczepanski/application-profiling

marcins committed:
Fix some missing shouldProfileApplication -> shouldTrace

devongovett committed:
Merge branch 'v2' into mszczepanski/application-profiling
